### PR TITLE
octopus: rgw: during GC defer, prevent new GC enqueue

### PR DESCRIPTION
**Pull request by ivancich**
backport tracker: https://tracker.ceph.com/issues/48331

---

backport of https://github.com/ceph/ceph/pull/38228
parent tracker: https://tracker.ceph.com/issues/47866

this backport was staged using ceph-backport.sh version 16.0.0.6848
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

**Comment by ivancich** [comment]
@smithfarm : this is a backport of bug fix that prevents an RGW data loss. Thus it's a very high priority. What's the process for getting this out to users ASAP?

**Comment by cbodley** [comment]
@ivancich i agree that we should expedite backports for this bug. but i don't think this pr is sufficient as a fix, because it will cause rados objects to leak when defer_gc() races with actual deletes. i think it's worth spending some more time on the fix to get it right

**Comment by mattbenjamin** [comment]
@cbodley Eric isn't proposing this as a fix;  However, it at worst trades an improper data loss for a space leak.  The former is much preferred, I think.

**Comment by mattbenjamin** [comment]
(@pritha-srivastava is working on a long-term fix for this issue;)
but @cbodley maybe I'm misunderstanding you;  an alternative would be, I think, to make deletes to proceed immediately, and allow any intersecting read(s) to fail when they continue past the delete point;  I think that should be a config option anyway.  Would that be preferable?

**Comment by ivancich** [comment]
> @cbodley Eric isn't proposing this as a fix; However, it at worst trades an improper data loss for a space leak. The former is much preferred, I think.

And it trades a much less frequent data leak for a much more frequent data loss. The data leak only happens if the this is a true defer (1. object read starts, 2. object removed, 3. object read continues) and that is going to be rare. And we have a tool -- rgw-orphan-list -- to help clean up.

On the other hand *any* sufficiently slow data read cause data to be deleted. And there's no undo for that.

If a more complete fix is put together and fully tested then sure. But I'd rather see this go out ASAP.

**Comment by smithfarm** [comment]
@ivancich Whether and when this makes it into an octopus release is up to @cbodley @mattbenjamin and @yuriw.

**Comment by cbodley** [comment]
@ivancich in irc discussion, @mattbenjamin and i agreed that disabling `defer_gc()` in radosgw is a better short-term workaround, since it fixes the issue without potential for space leaks

> (@pritha-srivastava is working on a long-term fix for this issue;)
> but @cbodley maybe I'm misunderstanding you; an alternative would be, I think, to make deletes to proceed immediately, and allow any intersecting read(s) to fail when they continue past the delete point; I think that should be a config option anyway. Would that be preferable? To clarify, I mean, just disable deferred delete period.

@mattbenjamin i think we're on the same page. to clarify with code, https://github.com/ceph/ceph/pull/38253 disables calls to `defer_gc()` in the read path, so long reads may fail if their tail objects go missing

i'm not sure it's worth making that configurable. ideally, the feature just works and nobody wants to turn it off. but either way, we should wait to add a config variable until we fix the feature, so it can be on by default from the start

**Comment by ivancich** [comment]
@smithfarm @yuriw So is it the case that Yuri decides when a version of Octopus is "cut" and released? In case I wasn't clear, that's what I'm getting at -- the stuff that happens after the PR approvals and even after the merge.

**Comment by smithfarm** [comment]
@ivancich No, it's the leads who decide that.

**Comment by cbodley** [comment]
i think it's worth backporting this commit along with the one in https://github.com/ceph/ceph/pull/38253. https://github.com/ceph/ceph/pull/38253 will prevent upgraded radosgws from making calls to `cls_rgw_gc_queue_update_entry()`, and this commit will prevent upgraded osds from adding new gc entries in the presence of un-upgraded radosgws

what do you think @ivancich? can you cherry-pick that commit here once it merges to master?

**Comment by smithfarm** [comment]
@ivancich One thing to keep in mind is that 15.2.6 (the most recent octopus release) was a so-called "hotfix" release, which means there is a relatively large number of already merged octopus PRs that are queued for 15.2.7.

So, the question becomes: do a "hotfix" 15.2.7 containing just this one fix, or do a normal 15.2.7 with all the already-merged fixed plus this fix?

**Review by cbodley — APPROVED** [review]

**Comment by ivancich** [comment]
I added the commit for https://github.com/ceph/ceph/pull/38253 to this backport.

**Comment by cbodley** [comment]
> So, the question becomes: do a "hotfix" 15.2.7 containing just this one fix, or do a normal 15.2.7 with all the already-merged fixes plus this fix?

thanks @smithfarm. that question was raised in the mailing list thread "Re: 15.2.7 (CORRECTED) QE Octopus validation status"

**Comment by mattbenjamin** [comment]
> 
> 
> @ivancich in irc discussion, @mattbenjamin and i agreed that disabling `defer_gc()` in radosgw is a better short-term workaround, since it fixes the issue without potential for space leaks
> 
> > (@pritha-srivastava is working on a long-term fix for this issue;)
> > but @cbodley maybe I'm misunderstanding you; an alternative would be, I think, to make deletes to proceed immediately, and allow any intersecting read(s) to fail when they continue past the delete point; I think that should be a config option anyway. Would that be preferable? To clarify, I mean, just disable deferred delete period.
> 
> @mattbenjamin i think we're on the same page. to clarify with code, #38253 disables calls to `defer_gc()` in the read path, so long reads may fail if their tail objects go missing
> 
> i'm not sure it's worth making that configurable. ideally, the feature just works and nobody wants to turn it off. but either way, we should wait to add a config variable until we fix the feature, so it can be on by default from the start

I think that it should probably be configurable in order to save the cost of deferred delete for sites that don't need/want it--clearly, not as a workaround for some future issue.  Agree on not making it configurable now, that would be part of the long-term fix, I think.

**Comment by ivancich** [comment]
Thank you for the quick handling, @yuriw!

**Pull request merged**
